app/db/item_list_store.py

```diff
@@ -0,0 +1,224 @@
+    to have been left over from a prior run.
+    """
+
+    RETRY_DELAY: ClassVar[float] = 1.0 * 60 * 15
+    """
+    Retries of failed item lists are delayed 15 minutes to let
+    the upstream systems recover from whatever their issue was. 
+    """
+
+    CLOCK_DRIFT: ClassVar[float] = 1.0 * 15
+    """
+    There are multiple participating clients of the Store and
+    we allow their clocks to drift by 15 seconds relative to
+    ours whenever we check the time on an item.
+    """
+
+    set_key: ClassVar[str] = redis.get_key("Item List Store")
+    """
+    The sorted set of items with their next-ready times.
+    """
+
+    item_list_channel_name: ClassVar[str] = redis.get_key("Item List Ready")
+    """
+    Redis pub/sub channel where items that are ready to process
+    are published so workers can pick them up.
+    """
+
+    db: Optional[Redis]
+
+    @classmethod
+    async def initialize(cls):
+        """
+        Make sure redis is connected and remember the connection pool.
+        """
+        await redis.connect_async()
+        cls.db = redis.db
+
+    @classmethod
+    async def terminate(cls):
+        """
+        Make sure redis is closed.
+        """
+        await redis.close_async()
+        cls.db = None
+
+    @classmethod
+    async def add_new_list(cls, key: str) -> Any:
+        """
+        Add a newly posted item list to the set with no delay.
+        Notify the item list channel that it's ready.
+        """
+        result = await cls.db.zadd(cls.set_key, score=now(), member=key)
+        cls.db.publish(cls.item_list_channel_name, key)
+        return result
+
+    @classmethod
+    async def add_retry_list(cls, key: str) -> Any:
+        """
+        Add a retry list to the set with an appropriate delay.
+        Also arrange to notify the item list channel when it's ready.
+        """
+
+        async def notify_later():
+            await asyncio.sleep(cls.RETRY_DELAY)
+            cls.db.publish(cls.item_list_channel_name, key)
+
+        result = await cls.db.zadd(
+            cls.set_key, score=now() + cls.RETRY_DELAY, member=key
+        )
+        asyncio.create_task(notify_later())
+        return result
+
+    @classmethod
+    async def remove_item_list(cls, key: str) -> Any:
+        """
+        Remove a processed item list from the set.
+        """
+        result = await cls.db.zrem(cls.set_key, key)
+        return result
+
+    @classmethod
+    async def select_from_channel(cls) -> Optional[str]:
+        """
+        Wait until there's an item sent to the channel, then return it.
+
+        This takes advantage of the fact that the underlying connection
+        pool has free connections and will automatically use one of them
+        for the subscription.  That way we can be a subscriber while
+        other tasks are also being publishers.
+        """
+        try:
+            channels = await cls.db.subscribe(cls.item_list_channel_name)
```

Comment:
Where are items being sent to this channel?

Reply:
In two places: one when they are posted by the web server (see `add_new_item`) and one when their retry delay has passed and they are ready to be processed again (see `add_retry_item`).